src/Common/src/TypeSystem/Interop/IL/MarshalHelpers.cs

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Internal.TypeSystem;
+using Internal.IL;
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.TypeSystem.Interop
+{
+    public static class MarshalHelpers
+    {
+        /// <summary>
+        /// Returns true if this is a type that doesn't require marshalling.
+        /// </summary>
+        public static bool IsBlittableType(TypeDesc type)
+        {
+            type = type.UnderlyingType;
+
+            if (type.IsValueType)
+            {
+                if (type.IsPrimitive)
+                {
+                    // All primitive types except char and bool are blittable
+                    TypeFlags category = type.Category;
+                    if (category == TypeFlags.Boolean || category == TypeFlags.Char)
+                        return false;
+
+                    return true;
+                }
+
+                foreach (FieldDesc field in type.GetFields())
+                {
+                    if (field.IsStatic)
+                        continue;
+
+                    TypeDesc fieldType = field.FieldType;
+
+                    // TODO: we should also reject fields that specify custom marshalling
+                    if (!MarshalHelpers.IsBlittableType(fieldType))
+                    {
+                        // This field can still be blittable if it's a Char and marshals as Unicode
```

Comment:
The correct way to implement this is to use whatever MarshalAs 'decoder' you have and cache the decision. You can also special case the few but it's error prone.